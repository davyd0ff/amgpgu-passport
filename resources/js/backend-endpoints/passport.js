const BACKEND_ENDPOINTS = {
  login: () => ({ method: 'post', url: 'api/login' }),
  refreshToken: () => ({ method: 'post', url: 'api/login/refresh' }),
  logout: () => ({ method: 'post', url: 'api/logout' }),

  getUserData: () => ({ method: 'get', url: 'api/user' }),
  getUserMenu: () => ({ method: 'get', url: 'api/user/menu' }),

  deleteFile: (fileId) => ({
    method: 'delete',
    url: `/api/files/delete/${fileId}`,
  }),
  fetchFiles: (context) => ({
    method: 'get',
    url: `/api/files/fetch/${context}`,
  }),
  uploadFile: (context) => ({
    method: 'post',
    url: `/api/files/upload/${context}`,
  }),

  getStudentData: () => ({ method: 'get', url: 'api/user/student-data' }),
  getStudentsTree: () => ({ method: 'get', url: 'api/students/tree' }),
  getStudentsTreeOfFaculty: (facultyCode) => ({
    method: 'get',
    url: `api/students/tree/${facultyCode}`,
  }),

  getIncomingNotifications: () => ({
    method: 'get',
    url: 'api/notifications/incoming',
  }),
  readNotification: (notificationId) => ({
    method: 'post',
    url: `api/notifications/read/${notificationId}`,
  }),
  readNotifications: () => ({
    method: 'post',
    url: 'api/notifications/read-all',
  }),

  addNotification: (context) => ({
    method: 'POST',
    url: `api/admin/notifications/add/for-${context}`,
  }),
};

export default BACKEND_ENDPOINTS;
