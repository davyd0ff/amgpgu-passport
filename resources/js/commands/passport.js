import PassportClient from '@/http/passportClient.js';
import BACKEND_ENDPOINTS from '@/backend-endpoints/passport.js';

function httpClient(config, options = {}) {
  const client = new PassportClient(options);
  return client.request(config);
}

const LoginCommands = {
  login: ({ username, password }) => {
    const backpoint = BACKEND_ENDPOINTS.login();
    const config = { ...backpoint, data: { username, password } };

    // todo think: "нарушается" сопряжение
    return httpClient(config, { enableProcessing: false });
  },
};

const StudentCommands = {
  getStudentData: () => {
    const backpoint = BACKEND_ENDPOINTS.getStudentData();
    return httpClient(backpoint);
  },
  getStudentsTree: (facultyCode = undefined) => {
    const backpoint = facultyCode
      ? BACKEND_ENDPOINTS.getStudentsTreeOfFaculty(facultyCode)
      : BACKEND_ENDPOINTS.getStudentsTree();
    return httpClient(backpoint);
  },
};

const FileCommands = {
  deleteFile: (fileId) => {
    const backpoint = BACKEND_ENDPOINTS.deleteFile(fileId);
    return httpClient(backpoint);
  },
  getFiles: (context) => {
    const backpoint = BACKEND_ENDPOINTS.getFiles(context);
    return httpClient(backpoint);
  },
  uploadFiles: (files, context, onUploadProgress) => {
    const backpoint = BACKEND_ENDPOINTS.uploadFile(context);

    let data = new FormData();
    files.forEach((file, index) => {
      data.append(`files[${index}]`.file, file.name);
    });
    // data.append(`files[0]`, file, file.name);

    const headers = {
      'Content-Type': 'multipart/form-data',
    };

    const options = { ...backpoint, headers, data, onUploadProgress };
    return httpClient(options);
  },
  uploadAvatar: (file) => {
    return this.uploadFile(file, 'avatar');
  },
};

const UserCommands = {
  getUserData: () => {
    const backpoint = BACKEND_ENDPOINTS.getUserData();
    return httpClient(backpoint);
  },
  getUserMenu: () => {
    const backpoint = BACKEND_ENDPOINTS.getUserMenu();
    return httpClient(backpoint);
  },
};

const NotificationCommands = {
  getNotifications: () => {
    const backpoint = BACKEND_ENDPOINTS.getNotifications();
    return httpClient(backpoint);
  },
  readNotification: (notificationId) => {
    const backpoint = BACKEND_ENDPOINTS.readNotification(notificationId);
    return httpClient(backpoint);
  },
  readNotifications: () => {
    const backpoint = BACKEND_ENDPOINTS.readNotifications();
    return httpClient(backpoint);
  },
};

export default {
  ...LoginCommands,
  ...UserCommands,
  ...StudentCommands,
  ...NotificationCommands,
  ...FileCommands,
};
