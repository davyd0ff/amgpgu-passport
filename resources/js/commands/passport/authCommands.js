import BACKEND_ENDPOINTS from '@/backend-endpoints/passport';
import httpClient from './httpClient';

export default {
  login: ({ username, password }) => {
    const backpoint = BACKEND_ENDPOINTS.login();
    const config = { ...backpoint, data: { username, password } };

    // todo think: "нарушается" сопряжение, данный метод знает какую функциональность несет в себе флаг
    return httpClient(config, { enableProcessing: false });
  },
  logout: () => {
    const backpoint = BACKEND_ENDPOINTS.logout();
    return httpClient(backpoint);
  },
};
