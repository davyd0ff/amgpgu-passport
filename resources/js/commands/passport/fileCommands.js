import BACKEND_ENDPOINTS from '@/backend-endpoints/passport';
import httpClient from './httpClient';

export default {
  deleteFile: (fileId) => {
    const backpoint = BACKEND_ENDPOINTS.deleteFile(fileId);
    return httpClient(backpoint);
  },
  getFiles: (context) => {
    const backpoint = BACKEND_ENDPOINTS.fetchFiles(context);
    return httpClient(backpoint).then((files) => Object.values(files));
  },
  uploadFiles: (files, context, onUploadProgress) => {
    const backpoint = BACKEND_ENDPOINTS.uploadFile(context);

    let data = new FormData();
    files.forEach((file, index) => {
      data.append(`files[${index}]`, file, file.name);
    });

    const headers = {
      'Content-Type': 'multipart/form-data',
    };

    const options = { ...backpoint, headers, data, onUploadProgress };
    return httpClient(options).then((data) => Object.values(data));
  },
  uploadAvatar: (file) => {
    return this.uploadFile(file, 'avatar');
  },
};
