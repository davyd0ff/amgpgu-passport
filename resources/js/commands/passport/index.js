import FileCommands from './fileCommands';
import AuthCommands from './authCommands';
import UserCommands from './userCommands';
import NotificationCommands from './notificationCommands';
import StudentCommands from './studentCommands';

export default {
  ...AuthCommands,
  ...UserCommands,
  ...StudentCommands,
  ...NotificationCommands,
  ...FileCommands,
};
