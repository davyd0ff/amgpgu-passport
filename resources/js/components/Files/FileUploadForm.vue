<template>
  <div>
    <v-file-input
      small-chips
      multiple
      label="Добавьте файлы"
      dense
      v-on:change="onChangeFiles"
    />
    <progressbar v-bind:progress="progress" />
    <div class="center-align">
      <a
        class="waves-effect btn teal"
        v-on:click="onUploadClick"
        v-bind:class="{ disabled: isUploadDisabled }"
      >
        <i class="material-icons right">cloud_upload</i>
        {{ 'FILE_UPLOAD_FORM_UPLOAD_BUTTON' | localize }}
      </a>
    </div>
  </div>
</template>

<script>
import Progressbar from '@/components/Progressbar';

export default {
  name: 'FileUploadForm',
  components: { Progressbar },
  props: {
    uploadFiles: { type: Function, default: () => Promise.resolve({}) },
  },
  data() {
    return {
      files: [],
      progress: undefined,
    };
  },
  computed: {
    isUploadDisabled() {
      return this.files.length === 0;
    },
  },
  methods: {
    onChangeFiles(files) {
      this.files = files;
    },
    onUploadClick() {
      this.uploadFiles(this.files, (progress) => {
        this.progress = Math.round((progress.loaded / progress.total) * 100);
      })
        .then(() => {
          this.files = [];
        })
        .catch(() => {});
    },
  },
};
</script>

<style scoped></style>
