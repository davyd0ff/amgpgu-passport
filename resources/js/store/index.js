import Vue from 'vue';
import Vuex from 'vuex';

import auth from './auth';
// import info from './info';
// import error from './error';
import student from './student';
import files from './files';
import user from './user';
import employee from './employee';
import notifications from './notifications';

Vue.use(Vuex);

export default new Vuex.Store({
  modules: {
    auth,
    user,
    student,
    files,
    employee,
    notifications,
  },
});
