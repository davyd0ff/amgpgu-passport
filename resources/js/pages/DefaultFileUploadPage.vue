<template>
  <div class="row">
    <div class="col s12">
      <div class="card">
        <div class="card-content">
          <file-upload-form v-bind:upload-files="onUpload" />

          <files-table v-bind:files="files" v-on:delete-file="onDelete" />
        </div>
      </div>
    </div>
  </div>
</template>

<script>
import FilesTable from '@/components/Files/FilesTable';
import FileUploadForm from '@/components/Files/FileUploadForm';
import passport from '@/commands/passport';

export default {
  name: 'DefaultFileUploadPage',
  components: { FileUploadForm, FilesTable },
  props: {
    context: String,
  },
  mounted() {
    this.onDownload();
  },
  computed: {
    files() {
      return this.$store.getters.files(this.context);
    },
  },
  methods: {
    onDownload() {
      this.$store
        .dispatch('getFiles', { context: this.context })
        .catch((error) => {
          this.$error(error);
        });
    },
    onUpload(filesToUpload, fnProgress) {
      return passport
        .uploadFiles(filesToUpload, this.context, fnProgress)
        .then((response) => {
          const uploadedfiles = Object.values(response.data);
          this.$store.dispatch('attachFiles', {
            context: this.context,
            files: uploadedfiles,
          });
        })
        .catch((error) => {
          this.$error(error);
          return Promise.reject(error);
        });
    },
    onDelete(fileId) {
      this.$store
        .dispatch('deleteFile', { fileId })
        .then(() => this.$message('DELETE_FILE_FILE_WAS_DELETED'))
        .catch((error) => this.$error(error));
    },
  },
};
</script>

<style scoped></style>
