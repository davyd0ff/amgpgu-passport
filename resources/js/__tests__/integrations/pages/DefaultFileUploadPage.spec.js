import { mount, createLocalVue } from '@vue/test-utils';
import Vuex from 'vuex';
import flushPromises from 'flush-promises';

import passport from '@/commands/passport';
import store from '@/store';
import DefaultFileUploadPage from '@/pages/DefaultFileUploadPage.vue';
import FileUploadForm from '@/components/Files/FileUploadForm.vue';

import { files } from '@/__tests__/__fixtures__/files';

jest.mock('@/commands/passport');

describe('DefaultFileUploadPage.vue', () => {
  const context = 'TEST';
  const $error = jest.fn();
  const $message = jest.fn();

  beforeEach(() => {
    passport.uploadFiles.mockClear();
    passport.getFiles.mockClear();
    $error.mockClear();
    $message.mockClear();
  });

  const makeWrapper = (options = {}) => {
    const localVue = createLocalVue();
    localVue.use(Vuex);
    localVue.filter('localize', (str) => str);

    return mount(DefaultFileUploadPage, {
      propsData: {
        context,
        ...(options.propsData ?? {}),
      },
      localVue,
      store,
      stubs: ['v-file-input'],
      mocks: {
        $error,
        $message,
      },
      ...options,
    });
  };

  it('render, when backend not returns files', async () => {
    passport.getFiles.mockImplementation(() => Promise.resolve([]));

    const wrapper = makeWrapper({});

    await flushPromises();
    expect(wrapper).toMatchSnapshot();
  });

  it('render, when backend returns error', async () => {
    passport.getFiles.mockImplementation(() =>
      Promise.reject({ code: 500, data: 'test' })
    );

    const wrapper = makeWrapper({});

    await flushPromises();
    expect($error).toHaveBeenCalledWith({
      code: 500,
      data: 'test',
    });
    expect(wrapper).toMatchSnapshot();
  });

  it('render, when backend returns files', async () => {
    passport.getFiles.mockImplementation(() => Promise.resolve(files));

    const wrapper = makeWrapper({});

    await flushPromises();
    expect(store.state.files[context]).toEqual(
      expect.arrayContaining([
        expect.objectContaining({
          id: expect.any(Number),
          name: expect.any(String),
          type: expect.any(String),
          created: expect.any(Number),
          url: expect.any(String),
        }),
      ])
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('uploadButton had been clicked and files uploaded', async () => {
    passport.uploadFiles.mockImplementation(() =>
      Promise.resolve([{ id: 'TEST_FILE_ID', name: 'TEST_FILE_NAME' }])
    );
    const wrapper = makeWrapper({});
    const childWrapper = wrapper.getComponent(FileUploadForm);

    await childWrapper.setData({ files: ['TEST_CHILD_FILE'] });
    await childWrapper.find('.btn').trigger('click');

    await flushPromises();
    expect(passport.uploadFiles).toHaveBeenCalledWith(
      ['TEST_CHILD_FILE'],
      context,
      expect.any(Function)
    );
    expect(store.state.files[context]).toEqual(
      expect.arrayContaining([
        expect.objectContaining({
          id: 'TEST_FILE_ID',
          name: 'TEST_FILE_NAME',
        }),
      ])
    );
    expect(childWrapper.vm.files).toStrictEqual([]);
  });

  it('uploadButton had been clicked and files did not upload', async () => {
    passport.uploadFiles.mockImplementation(() =>
      Promise.reject({ code: 401, data: 'test' })
    );
    const wrapper = makeWrapper();
    const childWrapper = wrapper.getComponent(FileUploadForm);

    await childWrapper.setData({ files: ['TEST_CHILD_FILE'] });
    await childWrapper.find('.btn').trigger('click');

    await flushPromises();
    expect(passport.uploadFiles).toHaveBeenCalledWith(
      ['TEST_CHILD_FILE'],
      context,
      expect.any(Function)
    );
    expect($error).toHaveBeenCalledWith({ code: 401, data: 'test' });
    expect(childWrapper.vm.files).toStrictEqual(['TEST_CHILD_FILE']);
  });

  it('delete a file, and the file has been deleted', async () => {
    passport.getFiles.mockImplementation(() => Promise.resolve(files));
    passport.deleteFile.mockImplementation(() =>
      Promise.resolve({ id: 'TEST' })
    );
    const wrapper = makeWrapper();

    wrapper.find('.btn-danger').trigger('click');

    await flushPromises();
    expect(passport.deleteFile).toHaveBeenCalledWith(expect.any(Number));
    expect($message).toHaveBeenCalledWith('DELETE_FILE_FILE_WAS_DELETED');
    expect($error).not.toHaveBeenCalled();
    expect(store.state.files[context]).toHaveLength(files.length - 1);
  });

  it('delete a file, and the file has not been deleted', async () => {
    passport.deleteFile.mockImplementation(() =>
      Promise.reject({ code: 401, data: 'test' })
    );
    const wrapper = makeWrapper();

    wrapper.find('.btn-danger').trigger('click');

    await flushPromises();
    expect(passport.deleteFile).toHaveBeenCalledWith(expect.any(Number));
    expect($message).not.toHaveBeenCalled();
    expect($error).toHaveBeenCalledWith({ code: 401, data: 'test' });
    expect(store.state.files[context]).toHaveLength(files.length);
  });
});
