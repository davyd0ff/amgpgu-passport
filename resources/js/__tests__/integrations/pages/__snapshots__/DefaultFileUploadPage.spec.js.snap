// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DefaultFileUploadPage.vue render, when backend not returns files 1`] = `
<div class="row">
  <div class="col s12">
    <div class="card">
      <div class="card-content">
        <div>
          <v-file-input-stub small-chips="" multiple="" label="Добавьте файлы" dense=""></v-file-input-stub>
          <!---->
          <div class="center-align"><a class="waves-effect btn teal disabled"><i class="material-icons right">cloud_upload</i>
              FILE_UPLOAD_FORM_UPLOAD_BUTTON
            </a></div>
        </div>
        <table>
          <thead>
            <tr>
              <th>#</th>
              <th>FILES_TABLE_COLUMN_TYPE_FILE</th>
              <th>FILES_TABLE_COLUMN_NAME_FILE</th>
              <th>FILES_TABLE_COLUMN_DATE_UPLOAD</th>
              <th>FILES_TABLE_COLUMN_DELETE_FILE</th>
            </tr>
          </thead>
          <tbody></tbody>
        </table>
        <!---->
      </div>
    </div>
  </div>
</div>
`;

exports[`DefaultFileUploadPage.vue render, when backend returns error 1`] = `
<div class="row">
  <div class="col s12">
    <div class="card">
      <div class="card-content">
        <div>
          <v-file-input-stub small-chips="" multiple="" label="Добавьте файлы" dense=""></v-file-input-stub>
          <!---->
          <div class="center-align"><a class="waves-effect btn teal disabled"><i class="material-icons right">cloud_upload</i>
              FILE_UPLOAD_FORM_UPLOAD_BUTTON
            </a></div>
        </div>
        <table>
          <thead>
            <tr>
              <th>#</th>
              <th>FILES_TABLE_COLUMN_TYPE_FILE</th>
              <th>FILES_TABLE_COLUMN_NAME_FILE</th>
              <th>FILES_TABLE_COLUMN_DATE_UPLOAD</th>
              <th>FILES_TABLE_COLUMN_DELETE_FILE</th>
            </tr>
          </thead>
          <tbody></tbody>
        </table>
        <!---->
      </div>
    </div>
  </div>
</div>
`;

exports[`DefaultFileUploadPage.vue render, when backend returns files 1`] = `
<div class="row">
  <div class="col s12">
    <div class="card">
      <div class="card-content">
        <div>
          <v-file-input-stub small-chips="" multiple="" label="Добавьте файлы" dense=""></v-file-input-stub>
          <!---->
          <div class="center-align"><a class="waves-effect btn teal disabled"><i class="material-icons right">cloud_upload</i>
              FILE_UPLOAD_FORM_UPLOAD_BUTTON
            </a></div>
        </div>
        <table>
          <thead>
            <tr>
              <th>#</th>
              <th>FILES_TABLE_COLUMN_TYPE_FILE</th>
              <th>FILES_TABLE_COLUMN_NAME_FILE</th>
              <th>FILES_TABLE_COLUMN_DATE_UPLOAD</th>
              <th>FILES_TABLE_COLUMN_DELETE_FILE</th>
            </tr>
          </thead>
          <tbody>
            <tr>
              <td>1</td>
              <td>application/octet-stream</td>
              <td><a href="https://passport.amgpgu.ru/storage/8hIOGo1pOVX78ttRoq8X01KxMSAyxwLlua2dnFZJ.bin" target="_blank">oauth-login.vsdx</a></td>
              <td>1616047842</td>
              <td><button class="btn btn-danger"><i class="medium material-icons">delete</i></button></td>
            </tr>
            <tr>
              <td>2</td>
              <td>application/vnd.openxmlformats-officedocument.wordprocessingml.document</td>
              <td><a href="https://passport.amgpgu.ru/storage/7ELSKEzm0RAz3Y1UAbzZa7dtYvdjwmAc4sGw5K8W.docx" target="_blank">Жалоба в ГИС ЖКХ.docx</a></td>
              <td>1616047843</td>
              <td><button class="btn btn-danger"><i class="medium material-icons">delete</i></button></td>
            </tr>
          </tbody>
        </table>
        <!---->
      </div>
    </div>
  </div>
</div>
`;
